### DISCUSSION TITLE
How to 'save' output of the `local_file_picker`

### DISCUSSION THREAD

**Discussion by tarensanders** [zauberzeug/nicegui#878] [Q&A] [answered]
### Question

I'm a bit of a newbie to async and nicegui. I'm using the code the the example for file picking, lightly edited so that it can also return a path to a folder.

I need two instances of the button to pick a file (an input dir and an output dir). This is simple enough:

```python
import os
from nicegui import ui
from local_file_picker import local_file_picker

default_outputdir = os.path.join(os.path.expanduser("~"), "Desktop")

async def pick_file() -> None:
    result = await local_file_picker("~", multiple=False)
    ui.notify(f"You chose {result}")

with ui.card():
    input_path = ui.label("")
    ui.button("Choose folder", on_click=pick_file).props("icon=folder flat size=sm")
    ui.markdown("**Output Folder**")
    output_path = ui.label(default_outputdir)
    out_button = ui.button("Change folder", on_click=pick_file).props(
        "icon=folder flat size=sm"
    )

ui.run()
```

I need to save and display the folder path. For one button this is simple - just update the text in `pick_file`

```python
async def pick_file() -> None:
    result = await local_file_picker("~", multiple=False)
    ui.notify(f"You chose {result}")
    input_path.text = result
```

But, I cannot work out how to generalise this to multiple buttons. I thought I could just add element as an argument to `pick_file` like so

```python
async def pick_file(element) -> None:
    result = await local_file_picker("~", multiple=False)
    ui.notify(f"You chose {result}")
    element.text = result

with ui.card():
    input_path = ui.label("")
    ui.button("Choose folder", on_click=lambda e: pick_file(input_path)).props("icon=folder flat size=sm")
    ui.markdown("**Output Folder**")
    output_path = ui.label(default_outputdir)
    out_button = ui.button("Change folder", on_click=lambda e: pick_file(output_path)).props(
        "icon=folder flat size=sm"
    )
```

But this gives an error `RuntimeWarning: coroutine 'pick_file' was never awaited`.

I'm sure there is a concept here I just don't understand. Any assistance appreciated!

**Comment by clueple** [comment]
I'm confused too ... don't really know how to get the path / a byte object from the ui.upload() .

Thanks for bringing up this (upload matter).

**Reply by falkoschindler** [reply]
Please read https://github.com/zauberzeug/nicegui/discussions/283 about why `ui.upload` doesn't return a path.

Here is an example of how to read the content of an uploaded file:
```py
def handle_upload(e: events.UploadEventArguments):
    text = e.content.read().decode()
    print(text)

ui.upload(on_upload=handle_upload)
```

Or in short:
```py
ui.upload(on_upload=lambda e: print(e.content.read().decode()))
```

**Reply by clueple** [reply]
Thanks for your hint.  Actually, I read the 'file upload' documentation, #283 and the local_file_picker example before.
But I can't find the 'on_upload' (only the 'on' ) and the events.UploadEventArugements.

I tried to dir(ui.upload()) to view the methods and attributes but I can't find the 'content' those kind of things.

Does it mean, the ui.upload() doesn't work for the local file? 

I'm just looking for an upload file solution to
1) read a csv file
2) store result of (1) into a pandas dataframe
3) create functions to add, delete, update row records of that (2) dataframe

I've been looking around for solutions.  It'd be nice if there's an example on such topic.

"_"

**Reply by falkoschindler** [reply]
The documentation for `ui.upload` explains [on_upload](https://nicegui.io/documentation/upload#file_upload) as well as all other [methods](https://nicegui.io/documentation/upload#methods) of this element.

As long as you don't need the _file path_, `ui.upload` is just fine for loading local or remote files. My example [above](https://github.com/zauberzeug/nicegui/discussions/878#discussioncomment-5789224) shows how to get the file content. In case of a CSV file you can create a dataframe and a `ui.table`. Alternatively, here is an example for editing data with a `ui.grid`: https://github.com/zauberzeug/nicegui/blob/main/examples/pandas_dataframe/main.py

**Reply by clueple** [reply]
> The documentation for `ui.upload` explains [on_upload](https://nicegui.io/documentation/upload#file_upload) as well as all other [methods](https://nicegui.io/documentation/upload#methods) of this element.
> 
> As long as you don't need the _file path_, `ui.upload` is just fine for loading local or remote files. My example [above](https://github.com/zauberzeug/nicegui/discussions/878#discussioncomment-5789224) shows how to get the file content. In case of a CSV file you can create a dataframe and a `ui.table`. Alternatively, here is an example for editing data with a `ui.grid`: https://github.com/zauberzeug/nicegui/blob/main/examples/pandas_dataframe/main.py

Thanks, @falkoschindler , I got it to upload CSV file(s) by 
1) turning the  upload_object.content into a Bytes object.
2) Then pandas read_csv that Bytes object in (1)
3) ui.aggrid(  the DataFrame object in (2) )

Thanks for your response.

**Comment by falkoschindler** [comment, marked as answer]
@tarensanders The problem is this:

A button accepts both synchronous and asynchronous event handlers. When passing `pick_file` to `on_click`, the button automatically awaits the async function.

But when wrapping it into a lambda statement, the `pick_file` function isn't awaited. Unfortunately there are no async lambdas in Python. Here is a minimal reproduction:
```py
async def sleep(duration: float):
    ui.notify('Sleeping...')
    await asyncio.sleep(duration)
    ui.notify('Done.')

ui.button('Click', on_click=lambda: sleep(1))
```

You have two options:
1. Use an async wrapper function without any arguments.
    ```py
    async def sleep(duration: float):
        ui.notify('Sleeping...')
        await asyncio.sleep(duration)
        ui.notify('Done.')
    
    async def sleep_for_1_second():
        await sleep(1)
    
    ui.button('Click', on_click=sleep_for_1_second)
    ```
2. Use `functools.partial`.
    ```py
    from functools import partial

    async def sleep(duration: float):
        ui.notify('Sleeping...')
        await asyncio.sleep(duration)
        ui.notify('Done.')
    
    ui.button('Click', on_click=partial(sleep, 1))
    ```

**Reply by tarensanders** [reply]
@falkoschindler that's a great explanation. Thank you. `functools.partial` works great.
